Modified ActorCriticPolicy to support non-shared features extractor

AlexPasqua opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
According to #1066 ([this comment](https://github.com/DLR-RM/stable-baselines3/issues/1066#issuecomment-1250030864) in particular).

## Description
<!--- Describe your changes in detail -->
Added a parameter `shared` to the `features_extractor_kwargs` (in the `policy_kwargs`).

If `shared` is false, the policy will create 2 instances of the same class of features extractor, called `pi_features_extractor` and `vf_features_extractor`.

The method `extract_features` has been overridden:
- in case `shared` is true, it simply calls the superclass' method;
- if it's false, it extracts the `pi_features` using the `pi_features_extractor` and the `vf_features` with the `vf_feratures_extractor`.

All other methods that were containing a features extraction (e.g. `forward`, `evaluate_actions`, `get_distribution`, `predict_values`) have been adapted to work with 2 sets of features in case `shared` is false: the `pi_features` and `vf_features` extracted by the method `extract_features` as just explained.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- You can use the syntax `closes #100` if this solves the issue #100 -->

According to #1066 ([this comment](https://github.com/DLR-RM/stable-baselines3/issues/1066#issuecomment-1250030864) in particular).

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

<!--- This Template is an edited version of the one from https://github.com/evilsocket/pwnagotchi/ -->

araffin commented on stable_baselines3/common/policies.py:
```diff
@@ -443,8 +443,17 @@ def __init__(
         self.activation_fn = activation_fn
         self.ortho_init = ortho_init
 
+        self.share_features_extractor = self.features_extractor_kwargs.get("shared", True)
```
I would rather add a parameter to the class, even though it means updating all classes.

AlexPasqua replied:
Alright! I had put it in the `features_extractor_kwargs` in order not to introduce too many modifications (and also because it's an argument related to the features extractor), but I'll put it in the class as you said 👌

araffin commented on stable_baselines3/common/policies.py:
```diff
@@ -617,14 +660,19 @@ def evaluate_actions(self, obs: th.Tensor, actions: th.Tensor) -> Tuple[th.Tenso
         # Preprocess the observation if needed
         features = self.extract_features(obs)
-        latent_pi, latent_vf = self.mlp_extractor(features)
+        if not self.share_features_extractor:
```
that's something I forgot but need to be mentioned in the doc: if the feature extractor is not shared, then you cannot share layers later on.

AlexPasqua replied:
Ok, I can add that to the doc!

Also, I could put a check after that line (670), something like "if shared_net not empty, raise error", but if, as you say in [this comment](https://github.com/DLR-RM/stable-baselines3/pull/1148#issuecomment-1299747968), you plan to remove the `shared_net` entirely, this might be useless.
Should I put the check in the meanwhile, and then remove it when the code gets refactored to remove the `shared_net`, in case you decide to do so?

AlexPasqua replied:
Ok I put it in the doc and I also put a check in the constructor of `ActorCriticPolicy`

araffin commented on stable_baselines3/common/policies.py:
```diff
@@ -639,17 +687,21 @@ def get_distribution(self, obs: th.Tensor) -> Distribution:
         :return: the action distribution.
         """
         features = self.extract_features(obs)
+        if not self.share_features_extractor:
```
I think it would be cleaner to add a `extract_actor_features()` and `extract_critic_features()` methods

AlexPasqua replied:
I agree, I'll do it!

araffin commented:
Thanks for the PR =)

Looking at the code and thinking for the future, it would probably make sense to always separate the networks after the feature extractor (so deprecate and remove the shared network in the mlp).
That would allow to have a cleaner distinction between actor and critic and have consistent code with the off-policy algorithms.

cc @ernestum @Miffyli @qgallouedec

AlexPasqua commented:
I've also modified the **doc**, but in doing this I removed the link to [issue #1066](https://github.com/DLR-RM/stable-baselines3/issues/1066#issuecomment-1246866844), because that guide would not be necessary anymore if we add this flag for shared/non-shared features extractor.
I don't know if you prefer to keep that link somewhere in the doc, in case I'll put it back.

Also, do you think the **tests** need to be updated for this features?

ernestum commented on stable_baselines3/common/policies.py:
```diff
@@ -547,11 +559,15 @@ def _build(self, lr_schedule: Schedule) -> None:
                 self.action_net: 0.01,
                 self.value_net: 1,
             }
+            if not self.share_features_extractor:
```
Removing the negation and switching the blocks might make this easier to read. Same in line 587 below.

AlexPasqua replied:
Did it

ernestum commented:
@araffin so you mean removing the ability to have any shared layers between value and policy network? That was always intended for reducing dimensionality or runtime but for me both never really improved anything. I don't know how it is for other applications though. If you are sufficiently confident that is never used anyways I am happy to remove the weight sharing feature to reduce the complexity of SB3.

araffin commented:
> o you mean removing the ability to have any shared layers between value and policy network? 

yes, I opened an issue here describing a tit bit more my motivation: https://github.com/DLR-RM/stable-baselines3/issues/1161

>  If you are sufficiently confident that is never used anyways I am happy to remove the weight sharing feature to reduce the complexity of SB3.

I would first deprecate and then remove.

araffin reviewed (commented):
I simplified the implementation and added a test, it's almost ready to be merged but the results are changed for `share_features_extractor=True` (I ran `python -m rl_zoo3.train --algo a2c --env PongNoFrameskip-v4 -n 10000 --seed 1619064180 -P` and check the values at 8000 steps).
I need to see if it's expected, for instance, if we changed the pseudo random generator state or if a bug was introduced.

EDIT: it's fine, it just comes from the order in `module_gains`

AlexPasqua commented on stable_baselines3/common/policies.py:
```diff
@@ -559,15 +563,18 @@ def _build(self, lr_schedule: Schedule) -> None:
             # features_extractor/mlp values are
             # originally from openai/baselines (default gains/init_scales).
             module_gains = {
+                self.features_extractor: np.sqrt(2),
```
The attributes `self.pi_features_extractor` and `self.vf_features_extractor` are created anyway in the `ActorCriticPolicy`'s constructor, and, in case the features extractor is shared, they are simply aliases for `self.features_extractor`.
So, at this point, wouldn't it be possible to do something like what follows?

```suggestion
                self.pi_features_extractor: np.sqrt(2),
```

araffin replied:
I removed `del self.features_extractor`, should be the cleaner solution

araffin commented:
@ernestum @qgallouedec could you have a look at it in the coming week?
the changes are relatively small but it's the start of a more major change, so I want to be sure we make it right ;)

qgallouedec commented on stable_baselines3/common/policies.py:
```diff
@@ -114,16 +114,17 @@ def make_features_extractor(self) -> BaseFeaturesExtractor:
-    def extract_features(self, obs: th.Tensor) -> th.Tensor:
+    def extract_features(self, obs: th.Tensor, features_extractor: Optional[BaseFeaturesExtractor] = None) -> th.Tensor:
         """
         Preprocess the observation if needed and extract features.
 
         :param obs:
         :return:
```
```suggestion
        Preprocess the observation if needed and extract features.

         :param obs: The observation
         :param features_extractor: The features extractor to use. If it is set to None, the features extractor of the policy is used.
         :return: The features

```

qgallouedec commented:
> That would allow to have a cleaner distinction between actor and critic and have consistent code with the off-policy algorithms.

I'm currently reviewing this PR and I agree. I think it would makes sense to simplify. I think it should be left for the future and that it is out of the scope of this PR, but I am willing to work on it.

qgallouedec commented on stable_baselines3/common/policies.py:
```diff
@@ -114,16 +114,18 @@ def make_features_extractor(self) -> BaseFeaturesExtractor:
         """
         Preprocess the observation if needed and extract features.
 
-        :param obs:
-        :return:
+         :param obs: The observation
+         :param features_extractor: The features extractor to use. If it is set to None, the features extractor of the policy is used.
+         :return: The features
         """
-        assert self.features_extractor is not None, "No features extractor was set"
+        features_extractor = features_extractor or self.features_extractor
+        assert features_extractor is not None, "No features extractor was set"
         preprocessed_obs = preprocess_obs(obs, self.observation_space, normalize_images=self.normalize_images)
-        return self.features_extractor(preprocessed_obs)
+        return features_extractor(preprocessed_obs)
```
I'm not very satisfied with this method, when you see it it's not easy to understand why in some cases you need to use the feature extractor of the instance, and in others not. I think that in the future it should be simplified.

araffin replied:
> I think that in the future it should be simplified.

how would you simplify it? 
always call it with a `features_extractor` parameter?

AlexPasqua replied:
> > I think that in the future it should be simplified.
> 
> how would you simplify it? always call it with a `features_extractor` parameter?

Yeah I personally think it could be a bit clearer to always pass the features extractor to that method, both in case it corresponds to `self.features_extractor` and in case it's a `pi` or `vf` features extractor.

I could do it, in case you opt for that

araffin replied:
@qgallouedec I added deprecation warning so we can make it the default in v1.8.0.

AlexPasqua commented:
> > That would allow to have a cleaner distinction between actor and critic and have consistent code with the off-policy algorithms.
> 
> I'm currently reviewing this PR and I agree. I think it would makes sense to simplify. I think it should be left for the future and that it is out of the scope of this PR, but I am willing to work on it.

I'd like to contribute too, if it's ok for you

qgallouedec commented:
> I'd like to contribute to this too, if it's ok for you

Of course it's ok! Thanks for your contributions @AlexPasqua!

AlexPasqua commented on tests/test_custom_policy.py:
```diff
@@ -28,17 +28,23 @@ def test_custom_offpolicy(model_class, net_arch):
 @pytest.mark.parametrize("optimizer_kwargs", [None, dict(weight_decay=0.0)])
 def test_custom_optimizer(model_class, optimizer_kwargs):
+    # Use different environment for DQN
+    if model_class is DQN:
+        env_id = "CartPole-v1"
+    else:
+        env_id = "Pendulum-v1"
```
I'd make this one line for a slightly better reading flow
```suggestion
    env_id = "CartPole-v1" if model_class is DQN else "Pendulum-v1"
```

araffin commented:
@AlexPasqua could you make a similar PR to SB3 contrib?

> Looking at the code and thinking for the future, it would probably make sense to always separate the networks after the feature extractor (so deprecate and remove the shared network in the mlp).
That would allow to have a cleaner distinction between actor and critic and have consistent code with the off-policy algorithms.

and for that, I would first do a PR that deprecates, then remove in a future major version

AlexPasqua commented:
> @AlexPasqua could you make a similar PR to SB3 contrib?

Sure! 👍 

> > Looking at the code and thinking for the future, it would probably make sense to always separate the networks after the feature extractor (so deprecate and remove the shared network in the mlp).
> > That would allow to have a cleaner distinction between actor and critic and have consistent code with the off-policy algorithms.
> 
> and for that, I would first do a PR that deprecates, then remove in a future major version

Alright! 👍

araffin commented:
@AlexPasqua unless you are already working on it (and do the PR today), I will have to do the SB3 contrib PR myself (we need a hotfix for ARS on SB3 contrib master)

AlexPasqua commented:
> @AlexPasqua unless you are already working on it (and do the PR today), I will have to do the SB3 contrib PR myself (we need a hotfix for ARS on SB3 contrib master)

Yes I'll do the PR today

araffin commented:
> Yes I'll do the PR today

thanks =)
in case you don't have time, tell me and I'll do it ;)

AlexPasqua commented:
Opened PR Stable-Baselines-Team
/
stable-baselines3-contrib#130

AlexPasqua commented:
> > Looking at the code and thinking for the future, it would probably make sense to always separate the networks after the feature extractor (so deprecate and remove the shared network in the mlp).
> > That would allow to have a cleaner distinction between actor and critic and have consistent code with the off-policy algorithms.
> 
> and for that, I would first do a PR that deprecates, then remove in a future major version

@araffin 
About the removal of the shared layers, should I start working on it already or should I wait that version 1.8 appears as WIP?

araffin commented:
> About the removal of the shared layers, should I start working on it already or should I wait that version 1.8 appears as WIP?

You could start working but I want to release v1.8.0 first (very soon, I will probably do it right atfer https://github.com/DLR-RM/stable-baselines3/pull/1260).
I also have some things specific in mind (I will probably help you for that):
- patch the loading (and move current patches to `save_util`)
- change the behavior of `net_arch=[64, 64]`
- try to provide meaningful error when someone try to load a shared net with SB3 v1.8.0

One thing that I'm not sure yet: shall we remove `MlpExtractor` completely? (and replace it with Actor/Critic class, the way it is done for off-policy algorithms)
I would say no for now.

AlexPasqua commented:
> * change the behavior of `net_arch=[64, 64]`

You mean making `net_arch=[64, 64]` equivalent to `net_arch=dict(pi=[64, 64], vf=[64, 64])`?

> One thing that I'm not sure yet: shall we remove `MlpExtractor` completely? (and replace it with Actor/Critic class, the way it is done for off-policy algorithms) I would say no for now.

I'd say no as well for now

araffin commented:
> You mean making net_arch=[64, 64] equivalent to net_arch=dict(pi=[64, 64], vf=[64, 64])?

yes! (to be consistent with SAC/TD3)
